Recipe service recipe controller recipe crud

amitgoldGH opened the pull request:
Add new recipe (POST), Update recipe (PUT), Read recipe (GET)

amitgoldGH committed:
Added putRecipeById and requestRecipeById (To update and to read recipes)

amitgoldGH committed:
Changed read return value to generic T

amitgoldGH committed:
Added validation check to create, Added read and update (with validation checks)

amitgoldGH committed:
Temporary fix.

amitgoldGH committed:
Added readREcipe and updateRecipe to service

amitgoldGH committed:
Added readRecipe and updateRecipe to RecipeServiceImplementation

amitgoldGH committed:
Correction on UserServiceImpl

amitgoldGH committed:
Added readRecipe and updateRecipe to RecipeServiceImplementation

RubyKozel commented on src/main/java/recipeapplication/controllers/RecipeController.java:
```diff
@@ -38,4 +34,21 @@ private Recipe toEntity(RecipeTO recipe) {
+						  @ModelAttribute("recipe") RecipeTO recipe) {
+		if (id.equalsIgnoreCase(recipe.getId()) &&
+				recipe.getCategory() != null &&
+				recipe.getIngridiant() != null &&
```
instead of checking for Null properties, you can use the annotation @Valid next to RecipeTO recipe and it would validat the properties (since each property have the @NotNull annotation)

amitgoldGH replied:
Has been edited to: (and is working)

public void recipePut(@RequestParam String id,
						   @RequestPart(value = "recipeImage", required = false) MultipartFile file,
						  @ModelAttribute("recipe") @Valid RecipeTO recipe ) {
		if (id.equalsIgnoreCase(recipe.getId()))
			this.recipes.updateRecipe(toEntity(recipe));
		}

RubyKozel replied:
Use @PathVariable("id") instead of @RequestParam for the id (the route should be "/recipe/{id}"

RubyKozel commented on src/main/java/recipeapplication/controllers/RecipeController.java:
```diff
@@ -22,7 +18,7 @@ public RecipeController(RecipeService recipes) {
 	}
 
-	@RequestMapping(value = "/recipe", method = RequestMethod.POST)
+	@RequestMapping(value = "/postRecipe", method = RequestMethod.POST)
```
leave the routes as in the design, the routes should be the same for every request, the distinguishement is made by the HTTP request itself

amitgoldGH replied:
Edited the routes for the POST and PUT methods to /recipe

@RequestMapping(value = "/recipe", method = RequestMethod.POST)
@RequestMapping(value = "/recipe", method = RequestMethod.PUT)

RubyKozel commented on src/main/java/recipeapplication/controllers/RecipeController.java:
```diff
@@ -38,4 +34,21 @@ private Recipe toEntity(RecipeTO recipe) {
+			this.recipes.updateRecipe(toEntity(recipe));
+		}
+
+	@RequestMapping(value = "/requestRecipeById", method = RequestMethod.GET)
```
/recipe/{id}

RubyKozel commented on src/main/java/recipeapplication/controllers/RecipeController.java:
```diff
@@ -38,4 +34,21 @@ private Recipe toEntity(RecipeTO recipe) {
+		}
+
+	@RequestMapping(value = "/requestRecipeById", method = RequestMethod.GET)
+	public Recipe recipeGet(@RequestParam String id)
```
@PathVariable("id")

amitgoldGH replied:
Changed to use @PathVariable("id") instead of @RequestParam

RubyKozel commented on src/main/java/recipeapplication/database/RecipeFirebaseCrud.java:
```diff
@@ -22,21 +26,73 @@ public RecipeFirebaseCrud(Firestore firestore,
+		} catch (InterruptedException e) {
+			e.printStackTrace();
+			return null;
+		} catch (ExecutionException e) {
```
its not. good behavior to return null as a response, lets throw a runtime exception if something is wrong. This way it will be easier for the client to know that something went wrong

amitgoldGH replied:
Changed to throw runtime exceptions.

RubyKozel replied:
@SagivAsraf wrote an exception handler for the User Controller, check that out and make a similar exception handler for the Recipe Controller

RubyKozel commented on src/main/java/recipeapplication/database/RecipeFirebaseCrud.java:
```diff
@@ -22,21 +26,73 @@ public RecipeFirebaseCrud(Firestore firestore,
+			else {
+				Map<String, Object> data = this.firestore.collection(collection).document(document).get().get().getData();
+
+				Recipe recipe = new Recipe();
```
Suggestion: export the creation of Recipe to a new method

amitgoldGH replied:
Changed to use the constructor instead

Recipe recipe = new Recipe(
						data.get("id").toString(),
						data.get("category").toString(),
						((List<String>) data.get("ingridiant")),
						data.get("preperation").toString(),
						data.get("recipeImageId") != null ? data.get("recipeImageId").toString() : null);

RubyKozel commented on src/main/java/recipeapplication/database/RecipeFirebaseCrud.java:
```diff
@@ -22,21 +26,73 @@ public RecipeFirebaseCrud(Firestore firestore,
+	public Recipe read(String document) {
+		System.out.println("In recipe reading section");
+		try {
+			DocumentSnapshot docSnap = this.firestore.collection(collection).document(document).get().get();
```
Suggestion: since you're using this more than once, consider exporting this to a new method

amitgoldGH replied:
Changed to using varibales to hold the references, no need to make a new method just to use ".get().get()" on the DocumentSnapshot variable.

RubyKozel commented on src/main/java/recipeapplication/database/RecipeFirebaseCrud.java:
```diff
@@ -22,21 +26,73 @@ public RecipeFirebaseCrud(Firestore firestore,
+			DocumentSnapshot docSnap = this.firestore.collection(collection).document(data.getId()).get().get();
+			if(docSnap.exists() && docSnap.get("id").toString().equalsIgnoreCase(data.getId())) {
+				this.firestore.collection(collection).document(data.getId()).set(data);
+			}
```
As discussed in WhatsApp, data shouldn't be set to easily when editing
First, you need to make sure no one is trying to change the ID
Second, some attributes might be null, as a user can decide to edit only few properties of the recipe, this set would overwrite the existing data in case some properties are null
Third, a user might want to edit the ingredients array and not entirely replace it, you should support merging the data coming from the user with the one already in the database. Check firestore documentation on how to merge data and edit data, maybe @SagivAsraf can help you here, since he's doing a similar thing in the UserService

RubyKozel commented on src/main/java/recipeapplication/services/RecipeServiceImpl.java:
```diff
@@ -43,4 +41,16 @@ public void addRecipe(Recipe recipe, MultipartFile recipeImage) {
+		if (Integer.parseInt(recipeId) > 0)
+			return this.recipes.read(recipeId);
+		else
+			return null;
```
As mentioned before, better to throw an exception than returning null values

amitgoldGH replied:
Has been edited to:

public Recipe readRecipe(String recipeId) {
		if (Integer.parseInt(recipeId) > 0)
			return this.recipes.read(recipeId);
		else
			throw new RuntimeException("Invalid id");
	}

amitgoldGH committed:
Changed requestmappings  and small fixes

amitgoldGH committed:
Changed null returns and add variable to hold ref

Changed to use variable to hold reference (docSnap) instead of calling this.firestore...

Changed null returns to throw RuntimeException

amitgoldGH committed:
changed null return to throw a RuntimeException

RubyKozel commented on src/main/java/recipeapplication/database/RecipeFirebaseCrud.java:
```diff
@@ -22,21 +28,58 @@ public RecipeFirebaseCrud(Firestore firestore,
+			else {
+				Map<String, Object> data = docSnap.getData();
+
+				return new Recipe(
```
Check how @SagivAsraf created the object using an ObjectMapper 
https://github.com/ronenrozen/recipeServer/pull/8/files (UserFirebaseCrud)

RubyKozel reviewed (changes requested):
Sync with @SagivAsraf about the changes he made in FirebaseCrud
You might need to rebase into master and fix your pull request after he merges his pull request

Also, let's discuss exception handling in the recipe controller after you merge this pull request

RubyKozel committed:
Merge branch 'master' into Recipe_Service_Recipe_Controller_Recipe_Crud

amitgoldGH committed:
Used objectmapper on read, added returns,
